spark_log_parser/loaders/__init__.py

```diff
@@ -0,0 +1,282 @@
+import abc
+import gzip
+import tarfile
+import logging
+
+from io import IOBase, BufferedIOBase
+from pathlib import Path
+from urllib.parse import ParseResult
+
+from aiodataloader import DataLoader
+from stream_unzip import stream_unzip
+
+FILE_SKIP_PATTERNS = [".DS_Store".lower(), "__MACOSX".lower(), "/."]
+
+
+class FileChunkStreamWrapper:
+    """
+    Minimal file-like wrapper to wrap the output of file streams that may be lacking methods like read/readlines (or in
+    other words, file streams that are raw generators). This allows us to wrap things like the output of `stream_unzip`
+    or botocore.StreamingBody.iter_lines() and pass those streams around as file-like objects.
+
+    Note - this wrapper does not implement the full file-like object API. Instead, it only implements at the moment
+        those methods that are actually called when loading files in various contexts. This doesn't mean that
+        new methods won't need to be implemented in the future as use-cases grow, it just means that at the moment
+        this minimal implementation satisfies the current use-cases we do have
+
+    Note #2 - once `chunks` is consumed, calls to read/lines will return empty byte buffers
+    """
+
+    _trailing_data: bytes = b''
+    _chunks = None
+
+    def __init__(self, chunks):
+        assert chunks
+        self._chunks = chunks
+
+    def __iter__(self):
+        """
+        The implementation of
+        """
+        return self.readlines()
+
+    def read(self, size=-1):
+        content = self._trailing_data
+        self._trailing_data = b''
+
+        if not size or size == -1:
+            for chunk in self._chunks:
+                content += chunk
+        else:
+            # iterate chunks until we've read at least `size` amount of data
+            for chunk in self._chunks:
+                content += chunk
+                if len(content) >= size:
+                    break
+
+            # Trim off any data that is past the provided `size` and hold on to it for subsequent `read` calls
+            if len(content) > size:
+                self._trailing_data = content[size:]
+                content = content[:size]
+
+        return content
+
+    def readlines(self):
+        """
+        Returns a stream of lines from self.chunks. This differs from many readlines() implementations, which tend to
+        read the whole file into memory and return a list of the lines
+        """
```

Comment:
Can we call it something else then since it doesn't meet the expectations of `readlines`? `iter_lines`?